### PR TITLE
[SPARK-35923][SQL] Coalesce empty partition with mixed CoalescedPartitionSpec and PartialReducerPartitionSpec

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Skip empty partitions in `ShufflePartitionsUtil.coalescePartitionsWithSkew`.

### Why are the changes needed?

Since [SPARK-35447](https://issues.apache.org/jira/browse/SPARK-35447), we apply `OptimizeSkewedJoin` before `CoalesceShufflePartitions`. However, There are something different with the order of these two rules.

Let's say if we have a skewed partitions: [0, 128MB, 0, 128MB, 0]:

* coalesce partitions first then optimize skewed partitions:
  [64MB, 64MB, 64MB, 64MB]
* optimize skewed partition first then coalesce partitions:
  [0, 64MB, 64MB, 0, 64MB, 64MB, 0]

So we can do coalesce in `ShufflePartitionsUtil.coalescePartitionsWithSkew` with mixed `CoalescedPartitionSpec` and `PartialReducerPartitionSpec` if `CoalescedPartitionSpec` is empty.

### Does this PR introduce _any_ user-facing change?

No, not release yet.

### How was this patch tested?

Add test.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @maropu @cloud-fan @viirya @JkSelf @yaooqinn

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala [hunk 1]
```diff
@@ -173,7 +176,11 @@ object ShufflePartitionsUtil extends Logging {
     if (numPartitions > start) {
       val partitionSpecs = coalescePartitions(
         partitionIndices(start), partitionIndices.last + 1, validMetrics, targetSize)
-      newPartitionSpecsSeq.zip(partitionSpecs).foreach(spec => spec._1 ++= spec._2)
+      assert(numPartitions > 0)
+      // skip empty partition iff all shuffle are empty after coalesced
+      if (!partitionSpecs.forall(specs => specs.size == 1 && specs.head.dataSize.contains(0))) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala:181]: I looked into `coalescePartitions`, it does skip empty partitions, but it needs to at least return one `CoalescedPartitionSpec`.

I think we can pass one more boolean flag to `coalescePartitions` to allow it to return no `CoalescedPartitionSpec`, for coalescing partitions between skewed partitions.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala]: SGTM !

**Review by yaooqinn — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala [hunk 2]
```diff
@@ -172,7 +172,7 @@ object ShufflePartitionsUtil extends Logging {
     // coalesce any partitions after the end of last skew section.
     if (numPartitions > start) {
       val partitionSpecs = coalescePartitions(
-        partitionIndices(start), partitionIndices.last + 1, validMetrics, targetSize)
+        partitionIndices(start), partitionIndices.last + 1, validMetrics, targetSize, true)
       newPartitionSpecsSeq.zip(partitionSpecs).foreach(spec => spec._1 ++= spec._2)
     }
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala:177]: How do we make sure we return at least one partition spec if there is no skew partition spec?

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala:177]: this code is under `coalescePartitionsWithSkew` which must have some skew partition specs

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
thank you @cloud-fan @yaooqinn

**Pull request closed without merging**